### ISSUE TITLE
insecure_open fails on Windows with a remote rabbitMQ server

### ISSUE THREAD

**Issue by zsx** [jgallagher/amiquip#8]
When tested with RabbitMQ on another computer with a slow-ish network connection, there's a good chance that insecure_open will fail.

Tested with the examples/hello_world_publish.rs with a remote RabbitMQ server. The backtrace is as below:
```
Error: Error { ctx: Os { code: 10057, kind: NotConnected, message: "A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied." }
stack backtrace:
   0: failure::backtrace::internal::InternalBacktrace::new (0x7ff7e1f40421)
             at C:\Users\szeng\.cargo\registry\src\github.com-1ecc6299db9ec823\failure-0.1.5\src\backtrace\internal.rs:44
   1: failure::backtrace::Backtrace::new (0x7ff7e1f3f43e)
             at C:\Users\szeng\.cargo\registry\src\github.com-1ecc6299db9ec823\failure-0.1.5\src\backtrace\mod.rs:111
   2: failure::error::error_impl::{{impl}}::from<std::io::error::Error> (0x7ff7e1eb6bd0)
             at C:\Users\szeng\.cargo\registry\src\github.com-1ecc6299db9ec823\failure-0.1.5\src\error\error_impl.rs:19
   3: failure::error::{{impl}}::from<std::io::error::Error> (0x7ff7e1e529ff)
             at C:\Users\szeng\.cargo\registry\src\github.com-1ecc6299db9ec823\failure-0.1.5\src\error\mod.rs:36
   4: core::convert::{{impl}}::into<std::io::error::Error,failure::error::Error> (0x7ff7e1e9531f)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libcore\convert.rs:501
   5: failure::context::Context<amiquip::errors::ErrorKind>::with_err<amiquip::errors::ErrorKind,std::io::error::Error> (0x7ff7e1e21dde)
             at C:\Users\szeng\.cargo\registry\src\github.com-1ecc6299db9ec823\failure-0.1.5\src\context.rs:105
   6: failure::Fail::context<std::io::error::Error,amiquip::errors::ErrorKind> (0x7ff7e1e9540a)
             at C:\Users\szeng\.cargo\registry\src\github.com-1ecc6299db9ec823\failure-0.1.5\src\lib.rs:161
   7: amiquip::frame_buffer::Inner<amiquip::frame_buffer::AmqpFrameKind>::read_from<amiquip::frame_buffer::AmqpFrameKind,mio::net::tcp::TcpStream,closure> (0x7ff7e1ede480)
             at C:\Users\szeng\work\amiquip.git\src\frame_buffer.rs:126
   8: amiquip::frame_buffer::FrameBuffer::read_from<mio::net::tcp::TcpStream,closure> (0x7ff7e1eddb66)
             at C:\Users\szeng\work\amiquip.git\src\frame_buffer.rs:24
   9: amiquip::io_loop::Inner::read_from_stream<mio::net::tcp::TcpStream,closure> (0x7ff7e1e8b2f8)
             at C:\Users\szeng\work\amiquip.git\src\io_loop\mod.rs:836
  10: amiquip::io_loop::IoLoop::handle_handshake_event<amiquip::auth::Auth,mio::net::tcp::TcpStream> (0x7ff7e1e7fac6)
             at C:\Users\szeng\work\amiquip.git\src\io_loop\mod.rs:370
  11: core::ops::function::FnMut::call_mut<fn(mut amiquip::io_loop::IoLoop*, mut mio::net::tcp::TcpStream*, mut amiquip::io_loop::handshake_state::HandshakeState<amiquip::auth::Auth>*, mio::event_imp::Event) -> core::result::Result<(), amiquip::errors::Error>,( (0x7ff7e1df7b95)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libcore\ops\function.rs:148
  12: amiquip::io_loop::IoLoop::run_io_loop<amiquip::io_loop::handshake_state::HandshakeState<amiquip::auth::Auth>,mio::net::tcp::TcpStream,fn(mut amiquip::io_loop::IoLoop*, mut mio::net::tcp::TcpStream*, mut amiquip::io_loop::handshake_state::HandshakeState<am (0x7ff7e1e836b0)
             at C:\Users\szeng\work\amiquip.git\src\io_loop\mod.rs:555
  13: amiquip::io_loop::IoLoop::run_amqp_handshake<amiquip::auth::Auth,mio::net::tcp::TcpStream> (0x7ff7e1e7ef84)
             at C:\Users\szeng\work\amiquip.git\src\io_loop\mod.rs:324
  14: amiquip::io_loop::IoLoop::thread_main<amiquip::auth::Auth,mio::net::tcp::TcpStream> (0x7ff7e1e7de2b)
             at C:\Users\szeng\work\amiquip.git\src\io_loop\mod.rs:308
  15: amiquip::io_loop::{{impl}}::start::{{closure}}<amiquip::auth::Auth,mio::net::tcp::TcpStream> (0x7ff7e1e7b2e6)
             at C:\Users\szeng\work\amiquip.git\src\io_loop\mod.rs:189
  16: std::sys_common::backtrace::__rust_begin_short_backtrace<closure,core::result::Result<(), amiquip::errors::Error>> (0x7ff7e1eb6606)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\sys_common\backtrace.rs:135
  17: std::thread::{{impl}}::spawn_unchecked::{{closure}}::{{closure}}<closure,core::result::Result<(), amiquip::errors::Error>> (0x7ff7e1e52526)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\thread\mod.rs:469
  18: std::panic::{{impl}}::call_once<core::result::Result<(), amiquip::errors::Error>,closure> (0x7ff7e1eb64c6)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\panic.rs:309
  19: std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure>,core::result::Result<(), amiquip::errors::Error>> (0x7ff7e1ee375c)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\panicking.rs:297
  20: panic_unwind::__rust_maybe_catch_panic (0x7ff7e22c4f62)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libpanic_unwind\lib.rs:87
  21: std::panicking::try<core::result::Result<(), amiquip::errors::Error>,std::panic::AssertUnwindSafe<closure>> (0x7ff7e1ee35a6)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\panicking.rs:276
  22: std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure>,core::result::Result<(), amiquip::errors::Error>> (0x7ff7e1eb6571)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\panic.rs:390
  23: std::thread::{{impl}}::spawn_unchecked::{{closure}}<closure,core::result::Result<(), amiquip::errors::Error>> (0x7ff7e1e52030)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\libstd\thread\mod.rs:468
  24: alloc::boxed::{{impl}}::call_box<(),closure> (0x7ff7e1e52813)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\src\liballoc\boxed.rs:749
  25: std::sys::windows::thread::{{impl}}::new::thread_start (0x7ff7e22c3a45)
             at /rustc/fc50f328b0353b285421b8ff5d4100966387a997\/src\libstd\sys\windows\thread.rs:47
  26: BaseThreadInitThunk (0x7fff54127974)
```

The error seems to say that the connection was not established yet, but the code should only run when the stream is readable (https://github.com/jgallagher/amiquip/blob/master/src/io_loop/mod.rs#L369), which is strange.

It worked fine when connecting to a local RabbitMQ server.

**Comment by jgallagher** [comment]
I agree that is very strange. Can you try setting the `RUST_LOG` environment variable to `trace` and post the dump of output?

**Comment by jgallagher** [comment]
I think this may be related to tokio-rs/mio#648. It sounds from that issue like on Windows, one can get spurious wakeups if you register for readable events before writing to the socket, and trying to read after one of those spurious events can fail with the NotConnected error you're seeing.

I just pushed an attempt at fixing this to the `windows-conn` branch. Can you try that branch out and see if it fixes your problem? If it does, I'll publish a point release. You should be able to grab the branch in your Cargo.toml via something like

```toml
[dependencies]
amiquip = { git = "https://github.com/jgallagher/amiquip", branch = "windows-conn" }
```

**Comment by zsx** [comment]
Thanks @jgallagher, I'll give it a try soon, but I just got this tracing (it's not easy to reproduce this when trace is on, it changed all of the timing, and I have to find a spot where WIFI is weak enough):
```
\amiquip.git> .\target\debug\examples\hello_world_publish.exe
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio_extras::timer] wakeup thread: sleep_until_tick=18446744073709551615; now_tick=0
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] register Token(65536) Readable | Writable
[2019-05-07T02:31:53Z TRACE mio_extras::timer] sleeping; tick_ms=100; now_tick=0; blocking sleep
[2019-05-07T02:31:53Z TRACE mio::sys::windows::tcp] scheduling a connect
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE amiquip::io_loop] reregistering socket for readable or writable
[2019-05-07T02:31:53Z TRACE mio::poll] registering with poller
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] reregister Token(65536) Readable | Writable
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] set readiness to (empty)
[2019-05-07T02:31:53Z TRACE mio::sys::windows::tcp] scheduling a read
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] set readiness to Readable
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] set readiness to Readable | Writable
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] select; timeout=Some(0ns)
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] polling IOCP
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] returning
[2019-05-07T02:31:53Z TRACE amiquip::io_loop] trying to write 8 bytes
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] set readiness to Readable
[2019-05-07T02:31:53Z TRACE mio::sys::windows::tcp] scheduling a write of 8 bytes
[2019-05-07T02:31:53Z TRACE mio::sys::windows::tcp] write error: A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied. (os error 10057)
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] set readiness to Readable | Writable
[2019-05-07T02:31:53Z TRACE amiquip::io_loop] wrote 8 bytes
[2019-05-07T02:31:53Z TRACE mio::sys::windows::selector] set readiness to Writable
[2019-05-07T02:31:53Z TRACE mio::sys::windows::tcp] scheduling a read
[2019-05-07T02:31:53Z TRACE mio::sys::windows::tcp] cancelling active TCP read
Error: Error { ctx: Os { code: 10057, kind: NotConnected, message: "A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied." }
```

**Comment by zsx** [comment]
That should have fixed it. I tried a few dozens of runs of the executables from master and windows-conn, I haven't seen any error from the one built from the 'windows-conn' branch, while I was still able to able reproduce from the other (just to make sure I was still at the right spot to reproduce the problem).

Thank you very much for the prompt fix.

**Comment by jgallagher** [comment]
Awesome! I'm relieved that worked. I just published `0.2.2` with this fix. I'll delete the `windows-conn` branch in a few days.
